### ISSUE TITLE
Found bug, have fix

### ISSUE THREAD

**Issue by spartacus777** [ArthurHub/Android-Image-Cropper#17]
If set OVAL shape, on some old androids (like 15, and 16 API I tested)
  canvas.clipPath(circleSelectionPath, Region.Op.XOR); work bad, we have  completely all image darken. I ggogled and found solution - we have to turn off hardware acceleration and it would work as it should. So, fix is

       if (Build.VERSION.SDK_INT <= 19 && mCropShape == CropImageView.CropShape.OVAL) {
            //TURN off hardware acceleration
            setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        }

Full code - 
  public void setCropShape(CropImageView.CropShape cropShape) {
        mCropShape = cropShape;

        if (Build.VERSION.SDK_INT <= 19 && mCropShape == CropImageView.CropShape.OVAL) {
            //TURN off hardware acceleration
            setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        }

        invalidate();
    }

AND in drawBackground method : 

        Path circleSelectionPath = new Path();
            mRectF.set(l, t, r, b);
            if (Build.VERSION.SDK_INT <= 19 && mCropShape == CropImageView.CropShape.OVAL) {
                mRectF.set(l + UIHelper.getPixel(2), t  + UIHelper.getPixel(2), r  - UIHelper.getPixel(2), b - UIHelper.getPixel(2));
            }
            circleSelectionPath.addOval(mRectF, Path.Direction.CW);
            canvas.save(Canvas.MATRIX_SAVE_FLAG);
            canvas.clipPath(circleSelectionPath, Region.Op.XOR);
            canvas.drawRect(bitmapRect.left, bitmapRect.top, bitmapRect.right, bitmapRect.bottom, mBackgroundPaint);
            canvas.restore();

Plese push this to repo
If you have any question, please contact me, I will help ^)
You have pretty good lib)

**Comment by ArthurHub** [comment]
If you create a pull request I will merge it.
Are you sure API 17-19 has this issue? 
Unfortunately I don't have the time to check it, can you make sure it is necessary up to API 19, I would really prefer not to disable hardware rendering if it is not necessary.
Thx for you time and effort, really appreciate it.

**Comment by spartacus777** [comment]
Not definetly sure, I will dot it while testing my whole app in a couple month. But from 5 android it works 100%, also I saw it worked on android 4.4, Samsung Tab 3

I can not create pull requestm since i intergated your code into my project and made a huge modifications to it (a lots of redesign work major)

But actually tou have jsut to add what i wrote above, and it should work

**Comment by Rainer-Lang** [comment]
Maybe your modifications helps the project...

**Comment by xitij** [comment]
Hi, I've noticed this issue as well when using the Oval crop tool. I get the following crash when attempting to load a photo from the gallery with a Oval crop tool:
```
E/AndroidRuntime: FATAL EXCEPTION: main
E/AndroidRuntime: java.lang.IllegalStateException: Underflow in restore - more restores than saves
E/AndroidRuntime:     at android.graphics.Canvas.native_restore(Native Method)
E/AndroidRuntime:     at android.graphics.Canvas.restore(Canvas.java:540)
E/AndroidRuntime:     at com.theartofdev.edmodo.cropper.cropwindow.CropOverlayView.drawBackground(CropOverlayView.java:558)
E/AndroidRuntime:     at com.theartofdev.edmodo.cropper.cropwindow.CropOverlayView.onDraw(CropOverlayView.java:314)
E/AndroidRuntime:     at android.view.View.draw(View.java:16178)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15174)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.draw(View.java:16181)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15174)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15169)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15169)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15169)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15169)
E/AndroidRuntime:     at android.view.View.draw(View.java:15948)
E/AndroidRuntime:     at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
E/AndroidRuntime:     at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
E/AndroidRuntime:     at android.view.View.draw(View.java:16181)
E/AndroidRuntime:     at com.android.internal.policy.PhoneWindow$DecorView.draw(PhoneWindow.java:2690)
E/AndroidRuntime:     at android.view.View.updateDisplayListIfDirty(View.java:15174)
E/AndroidRuntime:     at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:281)
E/AndroidRuntime:     at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:287)
E/AndroidRuntime:     at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:322)
E/AndroidRuntime:     at android.view.ViewRootImpl.draw(ViewRootImpl.java:2615)
E/AndroidRuntime:     at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2434)
E/AndroidRuntime:     at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
E/AndroidRuntime:     at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
E/AndroidRuntime:     at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
E/AndroidRuntime:     at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
E/AndroidRuntime:     at android.view.Choreographer.doCallbacks(Choreographer.java:670)
E/AndroidRuntime:     at android.view.Choreographer.doFrame(Choreographer.java:606)
E/AndroidRuntime:     at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
E/AndroidRuntime:     at android.os.Handler.handleCallback(Handler.java:739)
E/AndroidRuntime:     at android.os.Handler.dispatchMessage(Handler.java:95)
E/AndroidRuntime:     at android.os.Looper.loop(Looper.java:148)
E/AndroidRuntime:     at android.app.ActivityThread.main(ActivityThread.java:5417)
E/AndroidRuntime:     at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
E/AndroidRuntime:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```
The issue is in `CropOverlayView` in `drawBackground(...)`, `canvas.restore()` is called without a `canvas.save()`

I can submit a PR that fixes this if desired...

**Comment by ArthurHub** [comment]
finally added you fix in 1.0.7 version.
from my tests the fix is required up to API version 17 and not 19 (test 18 on emulator and 19 on nexus 5)
